File: backend/app/models/ui_element.py
"""
UI元素模型
"""
from sqlalchemy import Column, Integer, String, Text, JSON, DateTime, ForeignKey, Enum, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum
from app.core.database import Base


class LocatorType(str, enum.Enum):
    """定位策略枚举"""
    ID = "id"
    CSS = "css"
    XPATH = "xpath"
    TEXT = "text"
    LINK_TEXT = "link_text"
    PARTIAL_LINK_TEXT = "partial_link_text"
    TAG_NAME = "tag_name"
    NAME = "name"
    CLASS_NAME = "class_name"
    COMBINED = "combined"  # 组合定位


class ElementType(str, enum.Enum):
    """元素类型枚举"""
    BUTTON = "button"
    INPUT = "input"
    SELECT = "select"
    CHECKBOX = "checkbox"
    RADIO = "radio"
    LINK = "link"
    IMAGE = "image"
    TEXT = "text"
    DIV = "div"
    SPAN = "span"
    OTHER = "other"


class UIElement(Base):
    """UI元素模型"""
    __tablename__ = "ui_elements"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False, index=True)
    description = Column(Text)
    page_object_id = Column(Integer, ForeignKey("page_objects.id"), nullable=False)
    
    # 定位信息
    locator_type = Column(Enum(LocatorType), nullable=False)  # 定位策略
    locator_value = Column(String(500), nullable=False)  # 定位值
    locator_alternative = Column(JSON)  # 备用定位策略列表
    
    # 元素属性
    element_type = Column(Enum(ElementType))  # 元素类型
    is_required = Column(Boolean, default=False)  # 是否必填
    default_value = Column(String(500))  # 默认值
    
    # 等待策略
    wait_strategy = Column(JSON)  # 等待策略配置：显式等待、隐式等待、自定义等待条件
    
    # 操作封装
    operations = Column(JSON)  # 操作封装：点击、输入、选择等操作配置
    
    # 元数据
    tags = Column(JSON)  # 标签
    created_by = Column(Integer, ForeignKey("users.id"))
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # 关系
    page_object = relationship("PageObject", back_populates="elements")
    creator = relationship("User", backref="created_ui_elements")

File: frontend/src/store/services/uiElement.ts
import { api } from './api'

export interface UIElement {
  id: number
  name: string
  description?: string
  page_object_id: number
  locator_type: 'id' | 'css' | 'xpath' | 'text' | 'link_text' | 'partial_link_text' | 'tag_name' | 'name' | 'class_name' | 'combined'
  locator_value: string
  locator_alternative?: Array<Record<string, any>>
  element_type?: 'button' | 'input' | 'select' | 'checkbox' | 'radio' | 'link' | 'image' | 'text' | 'div' | 'span' | 'other'
  is_required?: boolean
  default_value?: string
  wait_strategy?: Record<string, any>
  operations?: Record<string, any>
  tags?: string[]
  created_by?: number
  created_at?: string
  updated_at?: string
}

export interface UIElementCreate {
  name: string
  description?: string
  page_object_id: number
  locator_type: 'id' | 'css' | 'xpath' | 'text' | 'link_text' | 'partial_link_text' | 'tag_name' | 'name' | 'class_name' | 'combined'
  locator_value: string
  locator_alternative?: Array<Record<string, any>>
  element_type?: 'button' | 'input' | 'select' | 'checkbox' | 'radio' | 'link' | 'image' | 'text' | 'div' | 'span' | 'other'
  is_required?: boolean
  default_value?: string
  wait_strategy?: Record<string, any>
  operations?: Record<string, any>
  tags?: string[]
}

export interface UIElementUpdate {
  name?: string
  description?: string
  page_object_id?: number
  locator_type?: 'id' | 'css' | 'xpath' | 'text' | 'link_text' | 'partial_link_text' | 'tag_name' | 'name' | 'class_name' | 'combined'
  locator_value?: string
  locator_alternative?: Array<Record<string, any>>
  element_type?: 'button' | 'input' | 'select' | 'checkbox' | 'radio' | 'link' | 'image' | 'text' | 'div' | 'span' | 'other'
  is_required?: boolean
  default_value?: string
  wait_strategy?: Record<string, any>
  operations?: Record<string, any>
  tags?: string[]
}

export interface UIElementListParams {
  page_object_id?: number
  locator_type?: string
  element_type?: string
  search?: string
  skip?: number
  limit?: number
}

export const uiElementService = {
  // 获取UI元素列表
  async getUIElements(params?: UIElementListParams): Promise<UIElement[]> {
    const response = await api.get<UIElement[]>('/ui-elements', { params })
    return response.data
  },

  // 获取UI元素详情
  async getUIElement(id: number): Promise<UIElement> {
    const response = await api.get<UIElement>(`/ui-elements/${id}`)
    return response.data
  },

  // 创建UI元素
  async createUIElement(data: UIElementCreate): Promise<UIElement> {
    const response = await api.post<UIElement>('/ui-elements', data)
    return response.data
  },

  // 更新UI元素
  async updateUIElement(id: number, data: UIElementUpdate): Promise<UIElement> {
    const response = await api.put<UIElement>(`/ui-elements/${id}`, data)
    return response.data
  },

  // 删除UI元素
  async deleteUIElement(id: number): Promise<void> {
    await api.delete(`/ui-elements/${id}`)
  },
}

